aten/src/ATen/native/native_functions.yaml

```diff
@@ -403,6 +403,10 @@
     SparseCsrCUDA: add_out_sparse_csr_cuda
     MkldnnCPU: mkldnn_add_out
 
+- func: add.quantized(Tensor self, Tensor other, float output_scale, int output_zero_point) -> Tensor
```

Comment:
So, my understanding is that there's already a manually-written dispatcher registration [here](https://github.com/pytorch/pytorch/blob/6adbe044e39c8e8db158d91e151aa6dead6e9aa4/aten/src/ATen/native/quantized/library.cpp#L23). And that there's an internal windows-only failure when you try to call the op? Is it called through something like this?
```
  return c10::Dispatcher::singleton()
      .findSchemaOrThrow("quantized::add", "")
      .typed<at::Tensor (const at::Tensor &, const at::Tensor &, double, int64_t)>()
      .call(self, other, output_scale, output_zero_point);
```

We probably don't want to expose this directly as a native function. I'm also not convinced that this would fix the bug, since this will expose a new `at::add(Tensor, Tensor, double, int64_t)` function for you to call, but it will invoke the same underlying dispatcher machinery.

If the above really doesn't work internally on windows, it's probably worth trying to fix it. Is there a more minimum repro- that you can show (hopefully that repro's in open source?).

Reply:
@bdhirsh 
Yes, I tried the option with Dispatcher.findSchemaOrThrow and it got the errors in windows builds: 
```
Summary: 
d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\impl\boxing.h(152): error C2066: cast to function type is illegal
d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\impl\boxing.h(152): error C2070: 'FuncType': illegal sizeof operand
d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\impl\boxing.h(152): error C2338: Function signature contains one or more unsupported parameter and/or return types. Look for a nearby error like "'call' is not a member of 'c10::impl::BoxedKernelWrapper<(your function type), void>'" - (your function type) is the unsupported signature.
d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\kernelfunction_impl.h(75): error C2039: 'call': is not a member of 'c10::impl::BoxedKernelWrapper<Return (at::Tensor,at::Tensor,double,int64_t),void>'
d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\kernelfunction_impl.h(61): error C2672: 'c10::KernelFunction::call': no matching overloaded function found
d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\kernelfunction_impl.h(61): error C2783: 'Return c10::KernelFunction::call(const c10::OperatorHandle &,c10::DispatchKeySet,Args...) const': could not deduce template argument for 'Return'
stderr: d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\impl\boxing.h(152): error C2066: cast to function type is illegal
d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\impl\boxing.h(152): error C2070: 'FuncType': illegal sizeof operand
d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\impl\boxing.h(152): error C2338: Function signature contains one or more unsupported parameter and/or return types. Look for a nearby error like "'call' is not a member of 'c10::impl::BoxedKernelWrapper<(your function type), void>'" - (your function type) is the unsupported signature.
d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\kernelfunction_impl.h(75): error C2039: 'call': is not a member of 'c10::impl::BoxedKernelWrapper<Return (at::Tensor,at::Tensor,double,int64_t),void>'
d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\kernelfunction_impl.h(61): error C2672: 'c10::KernelFunction::call': no matching overloaded function found
d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\kernelfunction_impl.h(61): error C2783: 'Return c10::KernelFunction::call(const c10::OperatorHandle &,c10::DispatchKeySet,Args...) const': could not deduce template argument for 'Return'
 ** Summary of failures encountered during the build **

Rule //arvr/projects/xrruntime/pc/Test/Test_WithHeadset:Test_WithHeadset_Package FAILED because Command failed with exit code 2.

command: [D:\ovrsource-null-hg\buck-out\gen\arvr\tools\buck\cl_tmp_override\cl_tmp_override, D:\ovrsource-null-hg\third-party\toolchains\vs2017_15.9\BuildTools\VC\Tools\MSVC\14.16.27023\bin\Hostx64\x64\cl.exe, @\\?\D:\ovrsource-null-hg\buck-out\bin\5ad17428\xplat\caffe2\torch_lib_ovrsource#compile-external_functions.cpp.obj802301fc,default\ppandcompile.argsfile]

stderr: d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\impl\boxing.h(152): error C2066: cast to function type is illegal

d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\kernelfunction_impl.h(61): note: see reference to class template instantiation 'c10::impl::BoxedKernelWrapper<Return (at::Tensor,at::Tensor,double,int64_t),void>' being compiled

        with

        [

            Return=at::Tensor

        ]

d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\dispatch\dispatcher.h(529): note: see reference to function template instantiation 'Return c10::KernelFunction::call<Return,at::Tensor,at::Tensor,double,int64_t>(const c10::OperatorHandle &,c10::DispatchKeySet,at::Tensor,at::Tensor,double,int64_t) const' being compiled

        with

        [

            Return=at::Tensor

        ]

d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\dispatch\dispatcher.h(412): note: see reference to function template instantiation 'Return c10::Dispatcher::call<Return,at::Tensor,at::Tensor,double,int64_t>(const c10::TypedOperatorHandle<at::Tensor (at::Tensor,at::Tensor,double,int64_t)> &,at::Tensor,at::Tensor,double,int64_t) const' being compiled

        with

        [

            Return=at::Tensor

        ]

d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\dispatch\dispatcher.h(412): note: while compiling class template member function 'Return c10::TypedOperatorHandle<at::Tensor (at::Tensor,at::Tensor,double,int64_t)>::call(at::Tensor,at::Tensor,double,int64_t) const'

        with

        [

            Return=at::Tensor

        ]

d:\ovrsource-null-hg\xplat\caffe2\torch\csrc\jit\tensorexpr\external_functions.cpp(128): note: see reference to function template instantiation 'Return c10::TypedOperatorHandle<at::Tensor (at::Tensor,at::Tensor,double,int64_t)>::call(at::Tensor,at::Tensor,double,int64_t) const' being compiled

        with

        [

            Return=at::Tensor

        ]

d:\ovrsource-null-hg\xplat\caffe2\torch\csrc\jit\tensorexpr\external_functions.cpp(127): note: see reference to class template instantiation 'c10::TypedOperatorHandle<at::Tensor (at::Tensor,at::Tensor,double,int64_t)>' being compiled

d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\ivalue_inl.h(289): note: see reference to class template instantiation 'c10::ArrayRef<T>' being compiled

        with

        [

            T=c10::IValue

        ]

d:\ovrsource-null-hg\xplat\caffe2\c10\core\memoryformat.h(63): note: see reference to class template instantiation 'c10::ArrayRef<int64_t>' being compiled

d:\ovrsource-null-hg\xplat\caffe2\c10\core\dispatchkeyset.h(57): note: see reference to class template instantiation 'std::initializer_list<c10::DispatchKey>' being compiled

d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\impl\boxing.h(152): error C2070: 'FuncType': illegal sizeof operand

        with

        [

            FuncType=at::Tensor (at::Tensor,at::Tensor,double,int64_t)

        ]

d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\impl\boxing.h(152): error C2338: Function signature contains one or more unsupported parameter and/or return types. Look for a nearby error like "'call' is not a member of 'c10::impl::BoxedKernelWrapper<(your function type), void>'" - (your function type) is the unsupported signature.

d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\kernelfunction_impl.h(75): error C2039: 'call': is not a member of 'c10::impl::BoxedKernelWrapper<Return (at::Tensor,at::Tensor,double,int64_t),void>'

        with

        [

            Return=at::Tensor

        ]

d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\kernelfunction_impl.h(75): note: see declaration of 'c10::impl::BoxedKernelWrapper<Return (at::Tensor,at::Tensor,double,int64_t),void>'

        with

        [

            Return=at::Tensor

        ]

d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\kernelfunction_impl.h(61): error C2672: 'c10::KernelFunction::call': no matching overloaded function found

d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\kernelfunction_impl.h(61): error C2783: 'Return c10::KernelFunction::call(const c10::OperatorHandle &,c10::DispatchKeySet,Args...) const': could not deduce template argument for 'Return'

d:\ovrsource-null-hg\xplat\caffe2\aten\src\aten\core\boxing\kernelfunction.h(156): note: see declaration of 'c10::KernelFunction::call'

    When running <c++ preprocess_and_compile>.

    When building rule //xplat/caffe2:torch_lib_ovrsource#compile-external_functions.cpp.obj802301fc,default (ovr_config//platform/windows:x86_64-ms14.16.27023-release-llvm-lld-vs15.9.12)..

Not all rules succeeded.
```

To fix those failures will be great!
The steps for reproduce:

Replace `auto r = at::native::quantized_add(qa, qb, out_qscale, out_qzero);` to call 
```
return c10::Dispatcher::singleton()
      .findSchemaOrThrow("quantized::add", "")
      .typed<at::Tensor (const at::Tensor &, const at::Tensor &, double, int64_t)>()
      .call(self, other, output_scale, output_zero_point);
```
and dry land it.

Reply:
Hmmm I'm not sure what's causing that windows error. But I still have a few questions:

(1) This PR shouldn't really fix this issue, because it doesn't change the logic in a meaningful way. If you build your PR locally and open up `build/aten/src/ATen/OperatorsEverything.cpp`, and grep for `add_quantized::call`, you'll see the code-generated logic that does something very similar to what I pasted above. Can you try debugging more and copying the codegen output manually and seeing if you still get the error?

(2) As another cut-through: do you need the dispatcher here? Or is there a reason you can't just directly make a call to `at::native::quantized_add()` at that call-site.

Reply:
@bdhirsh 
I can not call `at::native::quantized_add()` due to structure of our BUCK build (more info you can find at D32239783)

```
The ATen CPU library consists of two parts:
A. Common operator functions, e.g. "at::empty()", the list of sources can be found at "aten_cpu_source_list" in "tools/build_variables.bzl";
B. Implementations of these operators, e.g. "at::native::empty()", the list of sources is defined at "aten_native_source_list" in "tools/build_variables.bzl";
Note that A does not directly depend on B. A calls B via dispatch table. The dependency is injected into the dispatch table by B during its static initialization.
For internal mobile builds, B is built on a per-app basis. A is the public library for other libraries to depend on. Because these external functions call quantization functions that are not part of A, the NNC kernel library cannot resolve the missing symbols.

```

So we can not call at::native:: directly, only through DIspatcher